**One-bin 2A <-> 3B reaction, with 1st-order kinetics in both directions,
taken to equilibrium**

Diffusion not applicable (just 1 bin)

LAST REVISED: July 2, 2022

In [1]:
import set_path
set_path.add_ancestor_dir_to_syspath(3)  # The number of levels to go up 
                                         # to reach the project's home, from the folder containing this notebook

Added 'D:\Docs\- MY CODE\BioSimulations\life123-Win7' to sys.path


In [4]:
from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log

In [6]:
# Initialize the system
chem_data = chem(names=["A", "B"])     # NOTE: Diffusion not applicable (just 1 bin)

rxn = Reactions(chem_data)

# Reaction 2A <-> 3B , with 1st-order kinetics in both directions
rxn.add_reaction(reactants=[(2,"A")], products=[(3,"B")], forward_rate=5., reverse_rate=2.)

bio.initialize_system(n_bins=1, chem_data=chem_data, reactions=rxn)

bio.set_uniform_concentration(species_index=0, conc=10.)
bio.set_uniform_concentration(species_index=1, conc=50.)

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [10.]
  Species 1 (B). Diff rate: NOT SET. Conc:  [50.]


In [7]:
rxn.describe_reactions()

Number of reactions:  1
0: 2 A <-> 3 B  (Rf = 5.0 / Rb = 2.0)


In [8]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(2, 0, 1)] <-> [(3, 1, 1)]   ; Fwd: 5.0 / Back: 2.0


In [9]:
# First step
bio.react(time_step=0.1, n_steps=1)
bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [20.]
  Species 1 (B). Diff rate: NOT SET. Conc:  [35.]


_Early in the reaction :_
[A] = 20. [B] = 35.

In [10]:
# Numerous more steps
bio.react(time_step=0.1, n_steps=100)

bio.describe_state()

1 bins and 2 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [16.25]
  Species 1 (B). Diff rate: NOT SET. Conc:  [40.625]


Consistent with the 5/2 ratio of forward/reverse rates (and the 1st order reactions),
the systems settles in the following equilibrium:  [A] = 16.25 , [B] = 40.625

In [11]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
print(f"Ratio of equilibrium concentrations: {B_eq / A_eq}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations: 2.499999999999999
Ratio of forward/reverse rates: 2.5
